Added docs for background jobs (#59)
danielpaul/RapidRails: docs/background_jobs.md

# Background jobs

> Sidekiq is the background job processor we use that handles long-running tasks, such as sending emails or processing images, asynchronously.

Reviewer: link to sidekiq gem

> `AttachProfilePictureWorker` is a Sidekiq worker that downloads an image from a given URL and attaches it to the corresponding user record.

Reviewer: have a heading for jobs that we currently have in the platform already and this one being one  and what it does

Reviewer: also mention about how the emails by default are sent in the background if deliver_later is used and how sidekiq handles that

> 

Reviewer: mention about how on deployment they need to have another worker for the sidekiq for that process to run and sidekiq to process the queue. also need to mention how redis is required.

## Queue priority

Sidekiq fetches jobs from the queues in the order of their priority defined in the Sidekiq configuration file `config/sidekiq.yml`. The numbers next to the queue names represent the weight of the queues with a higher weight checked more often.

```
:queues:
  - [critical, 100]
  - [default, 50]
  - [mailers, 25]
  - [low_priority, 1]
```

## View queues

> Sidekiq also provides a web interface for monitoring job queues. If the necessary environment variables `SIDEKIQ_ADMIN_USERNAME` and `SIDEKIQ_ADMIN_PASSWORD` are set, it can be accessed at `localhost:3000/sidekiq`.

Reviewer: give instructions on how they can remove this if they don't want the web dashboard for this. 

say how they can customise the username and password for the login.

# Background jobs

[Sidekiq](https://github.com/sidekiq/sidekiq) is the background job processor we use that handles long-running tasks, such as sending emails or processing images, asynchronously.

## Configuration

For sidekiq to work and process jobs, you need to setup two things:

> 1. A worker `bundle exec sidekiq` needs to be added on your cloud platform and running. This command should be run on your local as well.

Reviewer: ```suggestion
1. A worker `bundle exec sidekiq` needs to run for the jobs to be processed. 
```

Reviewer: mention how if they are running the server with `bin/dev` then the process run automatically for local development.

> 2. Redis is also required. Set it up on your cloud platform as an add-on and set it up on your local with [this tutorial](https://redis.io/docs/getting-started/installation/install-redis-on-mac-os/).

Reviewer: ```suggestion
2. Redis for keeping track of the tasks queue. Set it up as a service on your servers and set it up on your local with [this tutorial](https://redis.io/docs/getting-started/installation/install-redis-on-mac-os/).
```

> ## List of default jobs

Reviewer: ```suggestion
## RaidRails Built-in Jobs
```

Sidekiq provides the [deliver_later](https://apidock.com/rails/v5.2.3/ActionMailer/MessageDelivery/deliver_later) method which when used with a **Mailer**, adds email jobs to the sidekiq queue. Other sidekiq jobs included in the codebase are:

1. [AttachProfilePictureWorker](../app/workers/attach_profile_picture_worker.rb) - downloads an image from a given URL and attaches it to the corresponding user record.
2. [UserAnonymizationWorker](../app/workers/user_anonymization_worker.rb) - runs the [AnonymizationService](../app/services/anonymization_service.rb) which anonymizes a discarded **User** record by clearing/randomizing all it's attributes.

## Queue priority

Sidekiq fetches jobs from the queues in the order of their priority defined in the Sidekiq configuration file [config/sidekiq.yml](../config/sidekiq.yml). The numbers next to the queue names represent the weight of the queues with a higher weight checked more often.

```
:queues:
  - [critical, 100]
  - [default, 50]
  - [mailers, 25]
  - [low_priority, 1]
```

## View queues

> Sidekiq also provides a web interface for monitoring job queues which can be accessed at `localhost:3000/sidekiq`

Reviewer: ```suggestion
Sidekiq also provides a web interface for monitoring job queues which can be accessed at `/sidekiq` on your local and your production server.
```

> To enable the web interface, set the username and password in the environment variables `SIDEKIQ_ADMIN_USERNAME` and `SIDEKIQ_ADMIN_PASSWORD` respectively.

Reviewer: ```suggestion
To enable the web interface, set the username and password in the environment variables `SIDEKIQ_ADMIN_USERNAME` and `SIDEKIQ_ADMIN_PASSWORD` in the Rails credentials file.
```

Author: > in the Rails credentials file.

these have been setup as environment variables, not credentials.

> To enable the web interface, set the username and password `SIDEKIQ_ADMIN_USERNAME` and `SIDEKIQ_ADMIN_PASSWORD` in the environment variables config.

Reviewer: ```suggestion
To enable the web interface, set the username and password `SIDEKIQ_ADMIN_USERNAME` and `SIDEKIQ_ADMIN_PASSWORD` in the environment variables.
```
